protected/components/StoredDatasetLinksPreview.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+
+class StoredDatasetLinksPreview extends DatasetComponents
+{
+    private $_id;
+    private $_db;
+//    private $_web;
+
+    public function __construct (int $id, CDbConnection $db_connection)
+    {
+        parent::__construct();
+        $this->_id = $id;
+        $this->_db = $db_connection;
+//        $this->_web = $webClient;
+    }
+
+    /**
+     * return the dataset ID
+     *
+     * @return int
+     */
+    public function getDatasetId(): int
+    {
+        return $this->_id;
+    }
+
+    /**
+     * return the dataset identifier (DOI)
+     *
+     * @return string
+     */
+    public function getDatasetDOI(): string
+    {
+        return $this->getDOIfromId($this->_db, $this->_id);
+    }
+
+    public function getImageUrl(): array
+    {
+        $sql = "select url from image where id = :id";
+
+        $command = $this->_db->createCommand($sql);
+        $command->bindParam( ":id", $this->_id , PDO::PARAM_INT);
+        $results = $command->queryAll();
+        return $results;
+    }
+
+    public function getPreviewDataForLinks(): array
+    {
+        $sql = "select identifier as short_doi, identifier as url, title, description, img.url as image_url from dataset dd, image img  where dd.id = img.id and dd.id = :id";
+        $command = $this->_db->createCommand($sql);
+        $command->bindParam( ":id", $this->_id , PDO::PARAM_INT);
+        $rows = $command->queryAll();
+        $results = [];
+
+        foreach ( $rows as $result) {
+            $result['url'] = "https://doi.org/10.5524/".$result['url'];
+            $results[]=$result;
+        }
+
+        return $results;
+    }
```

Comment:
@rija 

This function is implemented as described in #428, the `associative array` is returned as expected, and the `unit test` is passing. Would you please have a look for me?

Reply:
Hi @kencho51 

Remember what it says in my specifications:
>If a dataset has links of **"Additional information" type**, it will **retrieve** the meta-data associated with each of these links in order to construct a preview meta-data data structure.

The two words in bold are the most important aspects of the task's purpose and I didn't see any of them being used in your code for ``StoredDatasetLinksPreview.php``.

Then you will figure out what ``$webClient`` is for.

This diagram of the database schema may help you see how related links fit with dataset:
https://raw.githubusercontent.com/rija/gigadb-website/author-names-81/sql/gigadb_schema.svg

it's out-dated but the relevant parts  are still accurate.

Reply:
Hi @rija 

I have updated the `getPreviewDataForLinks` with `$webClient` included:
```
 public function getPreviewDataForLinks(): array
    {
        $sql = "select identifier as short_doi, exl.url as external_url, ext.name as type, title, description, img.url as image_url 
            from dataset dd, image img, external_link exl, external_link_type ext  
            where dd.id = img.id 
            and dd.id = exl.dataset_id 
            and exl.external_link_type_id = ext.id 
            and dd.id = :id;";
        $command = $this->_db->createCommand($sql);
        $command->bindParam( ":id", $this->_id , PDO::PARAM_INT);
        $rows = $command->queryAll();
        $results = [];

        foreach ( $rows as $result) {
            $response = null;
            try {
                $this->_web = new \GuzzleHttp\Client();
                $response = $this->_web->request('GET', $result['url']);
//                $response = $this->_web->request('GET', 'https://api.github.com/repos/guzzle/guzzle');
            }
            catch (RequestException $e) {
                Yii::log( Psr7\str($e->getRequest()) , "error");
                if ($e->hasResponse()) {
                    Yii::log( Psr7\str($e->getResponse()), "error");
                }
            }
            $result['response'] = $response !== null ? (string) $response->getBody() : null;
            $results[]=$result;
        }
        return $results;
    }
```
In this new implementation, **external_url** from `external_link` table and **type** from `external_link_type` table are included in returned array.

But for the `webClient`, what information would be needed to be stored from the `GET` response into the `array` for the preview purpose?

Here is the [example output](https://gist.github.com/kencho51/e13ca389321490032345ed08763fe089) of the `array`.